File: src/lib/ast/builder.test.ts
import { describe, it, expect } from 'vitest'
import {
  insertDigit, insertDecimalPoint, insertOperator,
  insertFraction, insertExponent, insertRadical, insertSquare,
  insertFunction, insertConstant, insertFactorial, insertNcr, insertNpr,
  insertParen, deleteCurrent, clearAll,
  insertCube, insertNthRadical, insertMixed, insertNumberLiteral,
} from './builder'
import { ASTNode, Cursor, INITIAL_CURSOR } from './types'

const C = (insertAt: number, path: Cursor['path'] = []): Cursor => ({ path, insertAt })

describe('insertDigit', () => {
  it('inserts NumberNode at empty expression', () => {
    const [nodes, cur] = insertDigit([], INITIAL_CURSOR, '3')
    expect(nodes).toEqual([{ type: 'number', value: '3' }])
    expect(cur).toEqual(C(1))
  })

  it('appends to adjacent NumberNode', () => {
    const init: ASTNode[] = [{ type: 'number', value: '3' }]
    const [nodes, cur] = insertDigit(init, C(1), '5')
    expect(nodes).toEqual([{ type: 'number', value: '35' }])
    expect(cur).toEqual(C(1))
  })

  it('inserts new NumberNode after operator', () => {
    const init: ASTNode[] = [{ type: 'operator', op: '+' }]
    const [nodes] = insertDigit(init, C(1), '5')
    expect(nodes[1]).toEqual({ type: 'number', value: '5' })
  })
})

describe('insertDecimalPoint', () => {
  it('appends dot to existing number', () => {
    const init: ASTNode[] = [{ type: 'number', value: '3' }]
    const [nodes] = insertDecimalPoint(init, C(1))
    expect((nodes[0] as any).value).toBe('3.')
  })

  it('creates 0. node when no prior number', () => {
    const [nodes] = insertDecimalPoint([], INITIAL_CURSOR)
    expect(nodes[0]).toEqual({ type: 'number', value: '0.' })
  })

  it('does not add second dot', () => {
    const init: ASTNode[] = [{ type: 'number', value: '3.' }]
    const [nodes] = insertDecimalPoint(init, C(1))
    expect((nodes[0] as any).value).toBe('3.')
  })
})

describe('insertOperator', () => {
  it('inserts operator node', () => {
    const [nodes] = insertOperator([], INITIAL_CURSOR, '+')
    expect(nodes[0]).toEqual({ type: 'operator', op: '+' })
  })
})

describe('insertFraction', () => {
  it('inserts fraction template and positions cursor in numerator', () => {
    const [nodes, cur] = insertFraction([], INITIAL_CURSOR)
    expect(nodes[0]).toMatchObject({ type: 'fraction', numerator: [], denominator: [] })
    expect(cur).toEqual({ path: [{ nodeIndex: 0, slot: 'numerator' }], insertAt: 0 })
  })
})

describe('insertExponent', () => {
  it('wraps previous node as base, cursor in exponent slot', () => {
    const init: ASTNode[] = [{ type: 'number', value: '2' }]
    const [nodes, cur] = insertExponent(init, C(1))
    expect(nodes[0]).toMatchObject({
      type: 'exponent',
      base: [{ type: 'number', value: '2' }],
      exponent: [],
    })
    expect(cur.path[0].slot).toBe('exponent')
    expect(cur.insertAt).toBe(0)
  })

  it('creates empty base when nothing before cursor', () => {
    const [nodes, cur] = insertExponent([], INITIAL_CURSOR)
    expect((nodes[0] as any).base).toEqual([])
    expect(cur.path[0].slot).toBe('exponent')
  })
})

describe('insertRadical', () => {
  it('inserts sqrt with empty degree, cursor in radicand', () => {
    const [nodes, cur] = insertRadical([], INITIAL_CURSOR, false)
    expect(nodes[0]).toMatchObject({ type: 'radical', degree: [], radicand: [] })
    expect(cur.path[0].slot).toBe('radicand')
  })

  it('inserts cbrt with degree=[3]', () => {
    const [nodes] = insertRadical([], INITIAL_CURSOR, true)
    expect((nodes[0] as any).degree).toEqual([{ type: 'number', value: '3' }])
  })
})

describe('insertSquare', () => {
  it('wraps previous node in exponent with exponent=[2]', () => {
    const init: ASTNode[] = [{ type: 'number', value: '4' }]
    const [nodes, cur] = insertSquare(init, C(1))
    expect(nodes[0]).toMatchObject({
      type: 'exponent',
      base: [{ type: 'number', value: '4' }],
      exponent: [{ type: 'number', value: '2' }],
    })
    expect(cur).toEqual(C(1))
  })
})

describe('insertFunction', () => {
  it('inserts function node, cursor in argument', () => {
    const [nodes, cur] = insertFunction([], INITIAL_CURSOR, 'sin')
    expect(nodes[0]).toMatchObject({ type: 'function', name: 'sin', argument: [] })
    expect(cur.path[0].slot).toBe('argument')
  })
})

describe('insertConstant', () => {
  it('inserts pi constant', () => {
    const [nodes] = insertConstant([], INITIAL_CURSOR, 'pi')
    expect(nodes[0]).toEqual({ type: 'constant', name: 'pi' })
  })
})

describe('insertFactorial', () => {
  it('wraps previous node in factorial', () => {
    const init: ASTNode[] = [{ type: 'number', value: '5' }]
    const [nodes, cur] = insertFactorial(init, C(1))
    expect(nodes[0]).toMatchObject({ type: 'factorial', operand: [{ type: 'number', value: '5' }] })
    expect(cur).toEqual(C(1))
  })
})

describe('insertNcr', () => {
  it('wraps previous node as n, cursor in r slot', () => {
    const init: ASTNode[] = [{ type: 'number', value: '5' }]
    const [nodes, cur] = insertNcr(init, C(1))
    expect(nodes[0]).toMatchObject({ type: 'ncr', n: [{ type: 'number', value: '5' }], r: [] })
    expect(cur.path[0].slot).toBe('r')
  })
})

describe('insertNpr', () => {
  it('wraps previous node as n, cursor in r slot', () => {
    const init: ASTNode[] = [{ type: 'number', value: '5' }]
    const [nodes, cur] = insertNpr(init, C(1))
    expect(nodes[0]).toMatchObject({ type: 'npr', n: [{ type: 'number', value: '5' }], r: [] })
    expect(cur.path[0].slot).toBe('r')
  })
})

describe('insertParen', () => {
  it('open: inserts paren node, cursor in children', () => {
    const [nodes, cur] = insertParen([], INITIAL_CURSOR, 'open')
    expect(nodes[0]).toMatchObject({ type: 'paren', children: [] })
    expect(cur.path[0].slot).toBe('children')
  })

  it('close: exits to parent level after paren node', () => {
    const init: ASTNode[] = [{ type: 'paren', children: [] }]
    const inner: Cursor = { path: [{ nodeIndex: 0, slot: 'children' }], insertAt: 0 }
    const [, cur] = insertParen(init, inner, 'close')
    expect(cur).toEqual(C(1))
  })
})

describe('deleteCurrent', () => {
  it('does nothing when expression empty at root', () => {
    const [nodes, cur] = deleteCurrent([], INITIAL_CURSOR)
    expect(nodes).toEqual([])
    expect(cur).toEqual(INITIAL_CURSOR)
  })

  it('removes last node', () => {
    const init: ASTNode[] = [{ type: 'number', value: '3' }, { type: 'operator', op: '+' }]
    const [nodes, cur] = deleteCurrent(init, C(2))
    expect(nodes).toEqual([{ type: 'number', value: '3' }])
    expect(cur).toEqual(C(1))
  })

  it('removes last digit from multi-digit number', () => {
    const init: ASTNode[] = [{ type: 'number', value: '35' }]
    const [nodes, cur] = deleteCurrent(init, C(1))
    expect((nodes[0] as any).value).toBe('3')
    expect(cur).toEqual(C(1))
  })

  it('exits to parent when at position 0 in a slot', () => {
    const init: ASTNode[] = [{ type: 'fraction', numerator: [], denominator: [] }]
    const inner: Cursor = { path: [{ nodeIndex: 0, slot: 'numerator' }], insertAt: 0 }
    const [, cur] = deleteCurrent(init, inner)
    expect(cur).toEqual(C(0))
  })
})

describe('clearAll', () => {
  it('returns empty expression and reset cursor', () => {
    const [nodes, cur] = clearAll()
    expect(nodes).toEqual([])
    expect(cur).toEqual(INITIAL_CURSOR)
  })
})

describe('insertCube', () => {
  it('wraps previous node as base with exponent 3', () => {
    const init: ASTNode[] = [{ type: 'number', value: '5' }]
    const [nodes] = insertCube(init, C(1))
    expect(nodes).toHaveLength(1)
    expect(nodes[0]).toMatchObject({ type: 'exponent', exponent: [{ type: 'number', value: '3' }] })
    expect((nodes[0] as any).base).toEqual([{ type: 'number', value: '5' }])
  })

  it('inserts empty-base exponent^3 when no previous node', () => {
    const [nodes] = insertCube([], INITIAL_CURSOR)
    expect(nodes[0]).toMatchObject({ type: 'exponent', base: [], exponent: [{ type: 'number', value: '3' }] })
  })
})

describe('insertNthRadical', () => {
  it('inserts radical with empty degree and cursor in degree slot', () => {
    const [nodes, cur] = insertNthRadical([], INITIAL_CURSOR)
    expect(nodes[0]).toMatchObject({ type: 'radical', degree: [], radicand: [] })
    expect(cur.path).toEqual([{ nodeIndex: 0, slot: 'degree' }])
    expect(cur.insertAt).toBe(0)
  })
})

describe('insertMixed', () => {
  it('inserts whole-number 0 then empty fraction, cursor starts in numerator', () => {
    const [nodes, cur] = insertMixed([], INITIAL_CURSOR)
    expect(nodes).toHaveLength(2)
    expect(nodes[0]).toMatchObject({ type: 'number', value: '0' })
    expect(nodes[1]).toMatchObject({ type: 'fraction', numerator: [], denominator: [] })
    expect(cur.path).toEqual([{ nodeIndex: 1, slot: 'numerator' }])
    expect(cur.insertAt).toBe(0)
  })
})

describe('insertNumberLiteral', () => {
  it('inserts a NumberNode with the given string value', () => {
    const [nodes, cur] = insertNumberLiteral([], INITIAL_CURSOR, '3.14')
    expect(nodes[0]).toEqual({ type: 'number', value: '3.14' })
    expect(cur.insertAt).toBe(1)
  })
})

File: test/core/ast/builder_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:scientific_calculator/core/ast/types.dart';
import 'package:scientific_calculator/core/ast/builder.dart';

Cursor c(int insertAt, [List<CursorSegment> path = const []]) =>
    Cursor(path, insertAt);

void main() {
  group('insertDigit', () {
    test('inserts NumberNode at empty expression', () {
      final (nodes, cur) = insertDigit([], initialCursor, '3');
      expect(nodes.length, 1);
      expect(nodes[0], isA<NumberNode>().having((n) => n.value, 'value', '3'));
      expect(cur.insertAt, 1);
      expect(cur.path, isEmpty);
    });

    test('appends to adjacent NumberNode', () {
      final init = [NumberNode('3')];
      final (nodes, cur) = insertDigit(init, c(1), '5');
      expect(nodes.length, 1);
      expect(nodes[0], isA<NumberNode>().having((n) => n.value, 'value', '35'));
      expect(cur.insertAt, 1);
      expect(cur.path, isEmpty);
    });

    test('inserts new NumberNode after operator', () {
      final init = [OperatorNode('+')];
      final (nodes, _) = insertDigit(init, c(1), '5');
      expect(nodes[1], isA<NumberNode>().having((n) => n.value, 'value', '5'));
    });
  });

  group('insertDecimalPoint', () {
    test('appends dot to existing number', () {
      final init = [NumberNode('3')];
      final (nodes, _) = insertDecimalPoint(init, c(1));
      expect((nodes[0] as NumberNode).value, '3.');
    });

    test('creates 0. node when no prior number', () {
      final (nodes, _) = insertDecimalPoint([], initialCursor);
      expect(nodes[0], isA<NumberNode>().having((n) => n.value, 'value', '0.'));
    });

    test('does not add second dot', () {
      final init = [NumberNode('3.')];
      final (nodes, _) = insertDecimalPoint(init, c(1));
      expect((nodes[0] as NumberNode).value, '3.');
    });
  });

  group('insertOperator', () {
    test('inserts operator node', () {
      final (nodes, _) = insertOperator([], initialCursor, '+');
      expect(nodes[0], isA<OperatorNode>().having((n) => n.op, 'op', '+'));
    });
  });

  group('insertFraction', () {
    test('inserts fraction template and positions cursor in numerator', () {
      final (nodes, cur) = insertFraction([], initialCursor);
      expect(nodes[0], isA<FractionNode>());
      final frac = nodes[0] as FractionNode;
      expect(frac.numerator, isEmpty);
      expect(frac.denominator, isEmpty);
      expect(cur.path.length, 1);
      expect(cur.path[0].nodeIndex, 0);
      expect(cur.path[0].slot, 'numerator');
      expect(cur.insertAt, 0);
    });
  });

  group('insertExponent', () {
    test('wraps previous node as base, cursor in exponent slot', () {
      final init = [NumberNode('2')];
      final (nodes, cur) = insertExponent(init, c(1));
      expect(nodes[0], isA<ExponentNode>());
      final exp = nodes[0] as ExponentNode;
      expect(exp.base.length, 1);
      expect(exp.base[0], isA<NumberNode>().having((n) => n.value, 'value', '2'));
      expect(exp.exponent, isEmpty);
      expect(cur.path[0].slot, 'exponent');
      expect(cur.insertAt, 0);
    });

    test('creates empty base when nothing before cursor', () {
      final (nodes, cur) = insertExponent([], initialCursor);
      expect((nodes[0] as ExponentNode).base, isEmpty);
      expect(cur.path[0].slot, 'exponent');
    });
  });

  group('insertRadical', () {
    test('inserts sqrt with empty degree, cursor in radicand', () {
      final (nodes, cur) = insertRadical([], initialCursor, false);
      expect(nodes[0], isA<RadicalNode>());
      final rad = nodes[0] as RadicalNode;
      expect(rad.degree, isEmpty);
      expect(rad.radicand, isEmpty);
      expect(cur.path[0].slot, 'radicand');
    });

    test('inserts cbrt with degree=[3]', () {
      final (nodes, _) = insertRadical([], initialCursor, true);
      final rad = nodes[0] as RadicalNode;
      expect(rad.degree.length, 1);
      expect(rad.degree[0], isA<NumberNode>().having((n) => n.value, 'value', '3'));
    });
  });

  group('insertSquare', () {
    test('wraps previous node in exponent with exponent=[2]', () {
      final init = [NumberNode('4')];
      final (nodes, cur) = insertSquare(init, c(1));
      expect(nodes[0], isA<ExponentNode>());
      final exp = nodes[0] as ExponentNode;
      expect(exp.base.length, 1);
      expect(exp.base[0], isA<NumberNode>().having((n) => n.value, 'value', '4'));
      expect(exp.exponent.length, 1);
      expect(exp.exponent[0], isA<NumberNode>().having((n) => n.value, 'value', '2'));
      expect(cur.insertAt, 1);
      expect(cur.path, isEmpty);
    });
  });

  group('insertFunction', () {
    test('inserts function node, cursor in argument', () {
      final (nodes, cur) = insertFunction([], initialCursor, FunctionName.sin);
      expect(nodes[0], isA<FunctionNode>());
      final fn = nodes[0] as FunctionNode;
      expect(fn.name, FunctionName.sin);
      expect(fn.argument, isEmpty);
      expect(cur.path[0].slot, 'argument');
    });
  });

  group('insertConstant', () {
    test('inserts pi constant', () {
      final (nodes, _) = insertConstant([], initialCursor, 'pi');
      expect(nodes[0], isA<ConstantNode>().having((n) => n.name, 'name', 'pi'));
    });
  });

  group('insertFactorial', () {
    test('wraps previous node in factorial', () {
      final init = [NumberNode('5')];
      final (nodes, cur) = insertFactorial(init, c(1));
      expect(nodes[0], isA<FactorialNode>());
      final fact = nodes[0] as FactorialNode;
      expect(fact.operand.length, 1);
      expect(fact.operand[0], isA<NumberNode>().having((n) => n.value, 'value', '5'));
      expect(cur.insertAt, 1);
      expect(cur.path, isEmpty);
    });
  });

  group('insertNcr', () {
    test('wraps previous node as n, cursor in r slot', () {
      final init = [NumberNode('5')];
      final (nodes, cur) = insertNcr(init, c(1));
      expect(nodes[0], isA<NcrNode>());
      final ncr = nodes[0] as NcrNode;
      expect(ncr.n.length, 1);
      expect(ncr.n[0], isA<NumberNode>().having((n) => n.value, 'value', '5'));
      expect(ncr.r, isEmpty);
      expect(cur.path[0].slot, 'r');
    });
  });

  group('insertNpr', () {
    test('wraps previous node as n, cursor in r slot', () {
      final init = [NumberNode('5')];
      final (nodes, cur) = insertNpr(init, c(1));
      expect(nodes[0], isA<NprNode>());
      final npr = nodes[0] as NprNode;
      expect(npr.n.length, 1);
      expect(npr.n[0], isA<NumberNode>().having((n) => n.value, 'value', '5'));
      expect(npr.r, isEmpty);
      expect(cur.path[0].slot, 'r');
    });
  });

  group('insertParen', () {
    test('open: inserts paren node, cursor in children', () {
      final (nodes, cur) = insertParen([], initialCursor, 'open');
      expect(nodes[0], isA<ParenNode>());
      expect((nodes[0] as ParenNode).children, isEmpty);
      expect(cur.path[0].slot, 'children');
    });

    test('close: exits to parent level after paren node', () {
      final init = [ParenNode([])];
      final inner = Cursor([CursorSegment(0, 'children')], 0);
      final (_, cur) = insertParen(init, inner, 'close');
      expect(cur.insertAt, 1);
      expect(cur.path, isEmpty);
    });
  });

  group('deleteCurrent', () {
    test('does nothing when expression empty at root', () {
      final (nodes, cur) = deleteCurrent([], initialCursor);
      expect(nodes, isEmpty);
      expect(cur.insertAt, 0);
      expect(cur.path, isEmpty);
    });

    test('removes last node', () {
      final init = [NumberNode('3'), OperatorNode('+')];
      final (nodes, cur) = deleteCurrent(init, c(2));
      expect(nodes.length, 1);
      expect(nodes[0], isA<NumberNode>().having((n) => n.value, 'value', '3'));
      expect(cur.insertAt, 1);
      expect(cur.path, isEmpty);
    });

    test('removes last digit from multi-digit number', () {
      final init = [NumberNode('35')];
      final (nodes, cur) = deleteCurrent(init, c(1));
      expect((nodes[0] as NumberNode).value, '3');
      expect(cur.insertAt, 1);
      expect(cur.path, isEmpty);
    });

    test('exits to parent when at position 0 in a slot', () {
      final init = [FractionNode([], [])];
      final inner = Cursor([CursorSegment(0, 'numerator')], 0);
      final (_, cur) = deleteCurrent(init, inner);
      expect(cur.insertAt, 0);
      expect(cur.path, isEmpty);
    });
  });

  group('clearAll', () {
    test('returns empty expression and reset cursor', () {
      final (nodes, cur) = clearAll();
      expect(nodes, isEmpty);
      expect(cur.insertAt, 0);
      expect(cur.path, isEmpty);
    });
  });

  group('insertCube', () {
    test('wraps previous node as base with exponent 3', () {
      final init = [NumberNode('5')];
      final (nodes, _) = insertCube(init, c(1));
      expect(nodes.length, 1);
      expect(nodes[0], isA<ExponentNode>());
      final exp = nodes[0] as ExponentNode;
      expect(exp.base.length, 1);
      expect(exp.base[0], isA<NumberNode>().having((n) => n.value, 'value', '5'));
      expect(exp.exponent.length, 1);
      expect(exp.exponent[0], isA<NumberNode>().having((n) => n.value, 'value', '3'));
    });

    test('inserts empty-base exponent^3 when no previous node', () {
      final (nodes, _) = insertCube([], initialCursor);
      expect(nodes[0], isA<ExponentNode>());
      final exp = nodes[0] as ExponentNode;
      expect(exp.base, isEmpty);
      expect(exp.exponent.length, 1);
      expect(exp.exponent[0], isA<NumberNode>().having((n) => n.value, 'value', '3'));
    });
  });

  group('insertNthRadical', () {
    test('inserts radical with empty degree and cursor in degree slot', () {
      final (nodes, cur) = insertNthRadical([], initialCursor);
      expect(nodes[0], isA<RadicalNode>());
      final rad = nodes[0] as RadicalNode;
      expect(rad.degree, isEmpty);
      expect(rad.radicand, isEmpty);
      expect(cur.path.length, 1);
      expect(cur.path[0].nodeIndex, 0);
      expect(cur.path[0].slot, 'degree');
      expect(cur.insertAt, 0);
    });
  });

  group('insertMixed', () {
    test('inserts whole-number 0 then empty fraction, cursor starts in numerator', () {
      final (nodes, cur) = insertMixed([], initialCursor);
      expect(nodes.length, 2);
      expect(nodes[0], isA<NumberNode>().having((n) => n.value, 'value', '0'));
      expect(nodes[1], isA<FractionNode>());
      final frac = nodes[1] as FractionNode;
      expect(frac.numerator, isEmpty);
      expect(frac.denominator, isEmpty);
      expect(cur.path.length, 1);
      expect(cur.path[0].nodeIndex, 1);
      expect(cur.path[0].slot, 'numerator');
      expect(cur.insertAt, 0);
    });
  });

  group('insertNumberLiteral', () {
    test('inserts a NumberNode with the given string value', () {
      final (nodes, cur) = insertNumberLiteral([], initialCursor, '3.14');
      expect(nodes[0], isA<NumberNode>().having((n) => n.value, 'value', '3.14'));
      expect(cur.insertAt, 1);
      expect(cur.path, isEmpty);
    });
  });
}
